lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2804,27 +2806,50 @@ votingEnabledInEra
     -> Either ErrConstructTx ()
 votingEnabledInEra = \case
     Write.RecentEraConway -> Right ()
-    _ -> Left ErrConstructTxVotingInWrongEra
+    _ -> Left $ ErrConstructTxVoting ErrWrongEra
 
 assertIsVoting
     :: DBLayer IO s
     -> Write.RecentEra era
     -> ExceptT ErrConstructTx IO ()
 assertIsVoting DBLayer{..} = \case
     Write.RecentEraConway -> do
-        voted <- liftIO $ atomically (alreadyVoted walletState)
-        if voted then
+        voted <- liftIO $ atomically (getCurrentVoting walletState)
+        if isJust voted then
             pure ()
         else
             throwE ErrConstructTxWithdrawalWithoutVoting
     _ -> pure ()
 
-alreadyVoted
+assertDifferentVoting
+    :: DBLayer IO s
+    -> Maybe VotingAction
+    -> Write.RecentEra era
+    -> ExceptT ErrConstructTx IO ()
+assertDifferentVoting DBLayer{..} votingActionM = \case
+    Write.RecentEraConway -> do
+        votingM <- liftIO $ atomically (getCurrentVoting walletState)
+        case votingM of
+            Nothing ->
+                pure ()
+            Just currentDrep -> case votingActionM of
+                Nothing ->
+                    pure ()
+                Just (VoteRegisteringKey requestedDrep) ->
+                    cmpDReps currentDrep requestedDrep
+                Just (Vote requestedDrep) ->
+                    cmpDReps currentDrep requestedDrep
+    _ -> pure ()
+  where
+    cmpDReps cDrep rDRep =
+        when (cDrep == rDRep) $ throwE $ ErrConstructTxVoting ErrWrongEra
```

Comment:
```suggestion
        liftIO (atomically $ getCurrentVoting walletState) >>= checkVotingIsDifferent votingActionM
    _ -> pure ()

checkVotingIsDifferent :: Monad m => Maybe VotingAction -> Maybe DRep -> ExceptT ErrConstructTx m ()
checkVotingIsDifferent votingActionM = \case
  Nothing ->
      pure ()
  Just currentDrep -> case votingActionM of
      Nothing ->
          pure ()
      Just vote -> checkDRepIsDifferent currentDrep (getDRep vote)
 where
    getDRep = \case
      VoteRegisteringKey requestedDrep -> requestedDrep
      Vote requestedDrep -> requestedDrep

    checkDRepIsDifferent cDrep rDRep =
        when (cDrep == rDRep) $ throwE $ ErrConstructTxVoting ErrWrongEra
```

Reply:
Seems to me cleaner as we separate the IO action of extracting some data from the DB with checking it. It could be good to make `checkVotingIsDifferent` pure and then move it somewhere else and unit test it but perhaps overkill for now.

Reply:
good suggestions, done. Thanks